liburing

jmillan opened the pull request:
Enable liburing usage for linux (kernel versions >= 6).

Use liburing to relay incoming RTP to Consumers. This greately reduces the number of system calls as just a single system call is needed to relay the incoming RTP packet to "every" Consumer, opposite to libuv which requires a system call to relay the RTP packet to "each" Consumer.

For efficiency, memory buffers are preallocated holding the data payload being sent.

- Building mediasoup in a Linux machine with kernel >= 6 will enable the feature.
- Running mediasoup in a Linux machine with kernel >= 6 will make use of the feature, if it was also enabled in build time.

NOTE: The numbers are really positive.  Not only the CPU usage is reduced but it's also more stable.

jmillan committed:
liburing

Enable liburing usage for linux (kernel versions >= 6).

Use liburing to relay incoming RTP to Consumers. This greately reduces the
number of system calls as just a single system call is needed to relay the
incoming RTP packet to "every" Consumer, opposite to libuv which requires a
system call to relay the RTP packet to "each" Consumer.

For efficiency, memory buffers are preallocated holding the data payload
being sent.

jmillan committed:
Update CHANGELOG.md

nazar-pc commented:
Okay, I'm glad you've tested it with positive outcome!
From my experience it is not always faster.

ibc commented:
> but it's also more stable

What does this mean exactly?

ibc commented on CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 
+### NEXT
+
+* Enable liburing usage for linux (kernel versions >= 6).
```
Add PR link now that it exists, please.

jmillan commented:
> Okay, I'm glad you've tested it with positive outcome!
> From my experience it is not always faster.

In my tests you'll see the difference when larg-ish Routers, but not with small/normal ones. Testing with >200 consumers you'll see a noticeable difference.

nazar-pc commented:
> In my tests you'll see the difference when larg-ish Routers, but not with small/normal ones. Testing with >200 consumers you'll see a noticeable difference.

That makes a lot of sense and is to be expected. Squeezing more performance out of the same CPU core is always nice.

jmillan committed:
add missing MS_LIBURING_ENABLED flag

jmillan committed:
Update CHANGELOG.md

jmillan commented:
>> The numbers are really positive. Not only the CPU usage is reduced but it's also more stable.

> What does this mean exactly?

The CPU usage is more stable, it shows less spikes.

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+	// Watch the event file descriptor.
+	this->uvHandle = new uv_poll_t;
+
+	error = uv_poll_init(DepLibUV::GetLoop(), uvHandle, efd);
```
```suggestion
	error = uv_poll_init(DepLibUV::GetLoop(), this->uvHandle, this->efd);
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+
+	if (error != 0)
+	{
+		delete uvHandle;
```
```suggestion
		delete this->uvHandle;
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+		MS_THROW_ERROR("uv_poll_init() failed: %s", uv_strerror(error));
+	}
+
+	uvHandle->data = this;
```
```suggestion
	this->uvHandle->data = this;
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+
+	uvHandle->data = this;
+
+	error = uv_poll_start(uvHandle, UV_READABLE, static_cast<uv_poll_cb>(onFdEvent));
```
```suggestion
	error = uv_poll_start(this->uvHandle, UV_READABLE, static_cast<uv_poll_cb>(onFdEvent));
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	// Initialize available UserData entries.
+	for (size_t i = 0; i < DepLibUring::QueueDepth; ++i)
```
```suggestion
	for (size_t i{ 0 }; i < DepLibUring::QueueDepth; ++i)
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+
+	if (this->efd < 0)
+	{
+		MS_THROW_ERROR("eventfd() failed: %s", std::strerror(-efd));
```
```suggestion
		MS_THROW_ERROR("eventfd() failed: %s", std::strerror(-this->efd));
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+
+	if (error >= 0)
+	{
+		MS_DEBUG_DEV("%i submission queue entries submitted", ret);
```
What is this `ret`? It's not declared here at all.

jmillan replied:
Fixed, should be `error` instead.

jmillan committed:
feedback

ibc commented on worker/src/handles/TcpConnectionHandle.cpp:
```diff
@@ -232,6 +233,26 @@ void TcpConnectionHandle::Write(
+			goto write_libuv;
+		}
+
+		auto prepared = DepLibUring::liburing->PrepareWrite(this->fd, data1, len1, data2, len2, cb);
```
Could you add some long comment here explaining why this could fail?

jmillan replied:
io_uring has a fixed capacity which is registered upon initialization. If the rings are in use completely at the moment of requesting, then we just fallback to libuv for this packet.

This should not happen as we are creating 1024  * 4 entries, meaning we should be able to send up to 1024 * 4 RTP packets in one shot. Usually a single Router has less than so many Consumers so this should not be a problem.

ibc commented on worker/src/handles/UdpSocketHandle.cpp:
```diff
@@ -157,6 +158,26 @@ void UdpSocketHandle::Send(
+			goto send_libuv;
+		}
+
+		auto prepared = DepLibUring::liburing->PrepareSend(this->fd, data, len, addr, cb);
```
Same here.

jmillan committed:
feedback

jmillan committed:
feedback, cont

jmillan committed:
feedback, cont

ibc commented on worker/include/DepLibUring.hpp:
```diff
@@ -0,0 +1,80 @@
+	}
+	io_uring* GetRing()
+	{
+		return &this->ring;
```
Let's use C++:

```suggestion
		return std::addressof(this->ring);
```

jmillan replied:
> Let's use C++:


hahaha

jmillan replied:
We need to rewrite mediasoup to C++!

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+{
+	MS_TRACE();
+
+	auto error = io_uring_queue_init(DepLibUring::QueueDepth, &this->ring, 0);
```
```suggestion
	auto error = io_uring_queue_init(DepLibUring::QueueDepth, std::addressof(this->ring), 0);
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+		MS_THROW_ERROR("eventfd() failed: %s", std::strerror(-this->efd));
+	}
+
+	error = io_uring_register_eventfd(&this->ring, this->efd);
```
```suggestion
	error = io_uring_register_eventfd(std::addressof(this->ring), this->efd);
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+	close(this->efd);
+
+	// Close the ring.
+	io_uring_queue_exit(&this->ring);
```
```suggestion
	io_uring_queue_exit(std::addressof(this->ring));
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+		return false;
+	}
+
+	auto* sqe = io_uring_get_sqe(&this->ring);
```
```suggestion
	auto* sqe = io_uring_get_sqe(std::addressof(this->ring));
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+		return false;
+	}
+
+	auto* sqe = io_uring_get_sqe(&this->ring);
```
```suggestion
	auto* sqe = io_uring_get_sqe(std::addressof(this->ring));
```

jmillan committed:
Add comments

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+	// Unset active flag.
+	this->active = false;
+
+	auto error = io_uring_submit(&this->ring);
```
```suggestion
	auto error = io_uring_submit(std::addressof(this->ring));
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,265 @@
+
+	this->availableUserDataEntries.pop();
+
+	auto* userData = &this->userDataBuffer[idx];
```
```suggestion
	auto* userData = std::addressof(this->userDataBuffer[idx]);
```

jmillan committed:
use std::addressoff() rather than &

jmillan committed:
fix typo

jmillan committed:
add comments

jmillan committed:
fix liburing Rust linkage

jmillan committed:
close the uv_poll_t handle

jmillan committed:
properly clean uv_poll_t handle

jmillan committed:
use 'err' rather than 'error' for consistency

jmillan committed:
enable liburing in runtime for kernel version >=6

The binary could have been prebuilt in a linux machine with kernel
version >= 6, but then such image could be run on an older kernel. We
need to check the kernel version in runtime and only enable the feature
when appropriate.

jmillan committed:
comment typo

jmillan committed:
disable clang-format for variable

jmillan committed:
new class LibUring

jmillan committed:
Do not create a LibUring instance if not supported by kernel

jmillan committed:
New line

jmillan committed:
meson: remove extra tabs

jmillan committed:
Merge remote-tracking branch 'origin/v3' into liburing

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -96,7 +94,7 @@ void DepLibUring::ClassInit()
 	// Enable liburing for kernel versions greather than or equal to 6.
 	if (kernelMayorLong >= 6)
```
I'd appreciate a static `IsSupported()` method in `LibUring` class instead.

ibc replied:
No.

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -96,7 +94,7 @@ void DepLibUring::ClassInit()
 	// Enable liburing for kernel versions greather than or equal to 6.
 	if (kernelMayorLong >= 6)
```
Please add a "info" tag log saying whether liburing is enabled or not.

jmillan committed:
handle feedback

jmillan committed:
IsSupported() is boolean

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,411 @@
+	int err = eventfd_read(liburing->GetEventFd(), std::addressof(v));
+	if (err < 0)
+	{
+		MS_ERROR("eventfd_read() failed: %s", std::strerror(-err));
```
If this error happens then what? There is no return; here. Code below weeks running.

jmillan replied:
If this error happens we should abort. This call should never fail.

ibc replied:
Good. Then change it to `MS_ABORT()`? ¯\_(ツ)_/¯

jmillan replied:
I'll modify it.

jmillan replied:
> Good. Then change it to MS_ABORT()? ¯_(ツ)_/¯

yes yes..., of course.

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,411 @@
+		MS_ERROR("eventfd_read() failed: %s", std::strerror(-err));
+	};
+
+	for (unsigned int i = 0; i < count; ++i)
```
`unsigned int i{ 0 }`

ibc commented on worker/src/RTC/Router.cpp:
```diff
@@ -660,6 +661,11 @@ namespace RTC
 			// Clone only happens if needed.
 			std::shared_ptr<RTC::RtpPacket> sharedPacket;
 
+#ifdef MS_LIBURING_ENABLED
+			// Activate liburing usage.
+			DepLibUring::SetActive();
+#endif
```
This is very confusing. Is it enabled because a macro/define is given? Or is it supported? And what does active mean?

jmillan replied:
`MS_LIBURING_ENABLED` means literaly => mediasoup-worker binary has liburing built in. This is, the capability is within the binary.

What would happen if a linux kernel >= 6 built the binary but a linux kernel < 6 is running it? Then even if the capability is within the binary, it cannot  be used in runtime. So `DepLibUring::SetActive()` will do nothing, indeed the thread local `LibUring` instance will be `nullptr`.

What would happen if a linux kernel >= 6 built the binary and a linux kernel >= 6 is running it? Then `DepLibUring::SetActive()` will effectively call `LibUring::SetActive`. We only send RTP packets while `LibUring` is active, and that's while the `Router` is iterating the `Consumers`, since we only want to use liburing for sending such packets for now.

jmillan replied:
@ibc, let me know if this is clear.

ibc replied:
So we use "enabled" to say that it's included in the binary and "supported" to say that it's effectively available in runtime? That's confusing, even more when we ALSO have "active", meaning that it's "enabled" in the binary, "supported" in runtime and in use right now.

ibc replied:
I would rename MS_LIBURING_ENABLED to MS_LIBURING_SUPPORTED and then have a static DepLibUring::IsSupported() that returns true if that macro is given and kernel version is good. Anyway, with my other proposal the macro is gonna be used just within DepLibUring hpp file (which now should be always included) and the rest of the code would just check that static method.

jmillan replied:
I'll rename `MS_LIBURING_ENABLED` by `MS_LIBURING_SUPPORTED` .

ibc reviewed (commented):
Please clarify in the code (add comments) about what enabled, supported and active is. AFAIU:

- If there is a macro/defined set by Meson, it should be about "supported" rather than "enabled".
- And then "active" means whether it is being used right now or not.
- I would completely avoid and get rid of "enabled" term.

ibc commented on worker/meson.build:
```diff
@@ -257,6 +257,28 @@ if host_machine.system() == 'windows'
+      'src/DepLibUring.cpp',
+    ]
+    cpp_args += [
+      '-DMS_LIBURING_ENABLED',
```
This should be `MS_LIBURING_SUPPORTED`.

ibc commented on worker/src/RTC/Router.cpp:
```diff
@@ -2,6 +2,7 @@
 // #define MS_LOG_DEV_LEVEL 3
 
 #include "RTC/Router.hpp"
+#include "DepLibUring.hpp"
```
I'm a bit worried about having to wrap every call to `DepLibUring` into a `#ifdef MS_LIBURING_SUPPORTED` block. IMHO `DepLibUring.hpp` should ALWAYS expose the same class and should only include `liburing.h` if the define is given. And it should expose a `IsSupported()` static method that returns true/false depending on that macro.

jmillan replied:
So you are asking to change:

```
#ifdef MS_LIBURING_SUPPORTED
```

by:

```
if (DepLibUring::IsSupported())
````

is that it?

ibc replied:
Yes. So only `DepLibUring::IsSupported()` method checks the `MS_LIBURING_SUPPORTED` given macro. Wdyt? Personally I don't like having all those `#if macro ... #end`. It's not readable. Or it's not real code.

jmillan replied:
The thing with the macro is that MacOS, Windows and binaries built with a linux kernel version <6 will not even have that symbol in the binary, and hence there is no code to run. Such code will not exist, simply. lf we put the macro behind a function everyone will call it even if liburing is not even present in the binary.

We are not going to burn the CPU for calling an empty method **every time** we send **anything** to the network.., but we can avoid making the call when the feature does not even exist in the binary by using a macro.

There are 3 places in the code making use of the macro: Router.cpp, UdpSocketHandler.cpp and TcpConnectionHanlder.cpp. Maybe we'll end up using liburing for every single thing we send over the network which will remove the usage in Router.cpp. It's two places where this macro is used.

EDIT: Also lib.cpp and Worker.cpp do use the macro to load the class.

ibc replied:
Here you are describing your original approach in this PR. Yes, I understand it. That's why I proposed another approach 😀. With my approach we must only check the macro in 3 places in DepLibUring.hpp (to only include liburing.h if supported, within the IsSupported() static method, and to only add that liburing struct/object member in the class). DepLibUring.hpp should always be included despite is Windows or Linux and DepLibUring.cpp should only be build if the macro is given (as it's already done in the PR).

Why don't you like this?

jmillan replied:
> Why don't you like this?

I don't find it bad using a macro not to get code compiled in architectures that is not going to be used. And as said in the previous comment, I don't find a reason to call a method in architectures and older linux machines that we know it will have no effect.

Any library out there has macros when it deals with multiple architectures and we are a library. This feature only works for linux.

Having said this, if given my rationale you insist on putting the macro inside a function that everyone will call then I will do it.

ibc replied:
> I don't find a reason to call a method in architectures and older linux machines that we know it will have no effect.

With my proposal the only method that would be called in NON supported platforms will be `DepLibUring::IsSupported()`.

> Having said this, if given my rationale you insist on putting the macro inside a function that everyone will call then I will do it.

No, I don't want to insist or mandate this change. I'm just exposing my concern. Let me check again how the code looks as it is.

jmillan committed:
handle feedback

jmillan committed:
feedback: Rename MS_LIBURING_ENABLED for MS_LIBURING_SUPPORTED

ibc commented on worker/include/DepLibUring.hpp:
```diff
@@ -0,0 +1,98 @@
+#ifndef MS_DEP_LIBURING_HPP
+#define MS_DEP_LIBURING_HPP
+
+#ifdef MS_LIBURING_SUPPORTED
```
I think it makes more sense this way: If not supported don't do absolutely anything.

```suggestion
#ifdef MS_LIBURING_SUPPORTED
#ifndef MS_DEP_LIBURING_HPP
#define MS_DEP_LIBURING_HPP

```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,411 @@
+#define MS_CLASS "DepLibUring"
+
+// #define MS_LOG_DEV_LEVEL 3
```
For consistency:

```suggestion
#define MS_CLASS "DepLibUring"
// #define MS_LOG_DEV_LEVEL 3
```

ibc commented on worker/src/RTC/Router.cpp:
```diff
@@ -2,6 +2,7 @@
 // #define MS_LOG_DEV_LEVEL 3
 
 #include "RTC/Router.hpp"
+#include "DepLibUring.hpp"
```
So if we check the macro we must do it everywhere. We do it in Worker.cpp so must also be done here:

```suggestion
#ifdef MS_LIBURING_SUPPORTED
#include "DepLibUring.hpp"
#endif
```

ibc commented on worker/include/DepLibUring.hpp:
```diff
@@ -0,0 +1,98 @@
+			this->active = true;
+		}
+		bool IsActive() const
+		{
+			return this->active;
+		}
+		io_uring* GetRing()
+		{
+			return std::addressof(this->ring);
+		}
+		int GetEventFd() const
+		{
+			return this->efd;
+		}
+		void ReleaseUserDataEntry(size_t idx)
+		{
+			this->availableUserDataEntries.push(idx);
+		}
+
+	private:
+		UserData* GetUserData();
+
+	private:
+		// io_uring instance.
+		io_uring ring;
+		// Event file descriptor to watch for completions.
+		int efd;
+		// libuv handle used to poll uring completions.
+		uv_poll_t* uvHandle{ nullptr };
+		// Whether we are currently sending RTP over io_uring.
+		bool active{ false };
+		// Pre-allocated UserData entries.
+		UserData userDataBuffer[QueueDepth]{};
+		// Indexes of available UserData entries.
+		std::queue<size_t> availableUserDataEntries;
+	};
+};
+
+#endif
+#endif
```
So let's not check the macro here. Instead, the rest of the code should NOT include `DepLibUring.hpp` if the macro is not given. Consistency with your own proposal:
- We only include `DepLibUring.hpp` if the macro is given.
- We only call methods in `DepLibUring` if the macro is given.
- Everywhere.
- We don't add any macro check in `DepLibUring.hpp` or `DepLibUring.cpp` since those files should ONLY be included and built if the macro is given.

```suggestion
#ifndef MS_DEP_LIBURING_HPP
#define MS_DEP_LIBURING_HPP

#include "DepLibUV.hpp"
#include <functional>
#include <liburing.h>
#include <queue>

class DepLibUring
{
public:
	using onSendCallback = const std::function<void(bool sent)>;

	/* Struct for the user data field of SQE and CQE. */
	struct UserData
	{
		uint8_t store[1500]{};
		onSendCallback* cb{ nullptr };
		size_t idx{ 0 };
	};

	/* Number of submission queue entries (SQE). */
	static constexpr size_t QueueDepth{ 1024 * 4 };

	static bool IsSupported();
	static void ClassInit();
	static void ClassDestroy();
	static void StartPollingCQEs();
	static void StopPollingCQEs();
	static bool PrepareSend(
	  int sockfd, const void* data, size_t len, const struct sockaddr* addr, onSendCallback* cb);
	static bool PrepareWrite(
	  int sockfd, const void* data1, size_t len1, const void* data2, size_t len2, onSendCallback* cb);
	static void Submit();
	static void SetActive();
	static bool IsActive();

	class LibUring;

	thread_local static LibUring* liburing;

public:
	class LibUring
	{
	public:
		LibUring();
		~LibUring();
		void StartPollingCQEs();
		void StopPollingCQEs();
		bool PrepareSend(
		  int sockfd, const void* data, size_t len, const struct sockaddr* addr, onSendCallback* cb);
		bool PrepareWrite(
		  int sockfd, const void* data1, size_t len1, const void* data2, size_t len2, onSendCallback* cb);
		void Submit();
		void SetActive()
		{
			this->active = true;
		}
		bool IsActive() const
		{
			return this->active;
		}
		io_uring* GetRing()
		{
			return std::addressof(this->ring);
		}
		int GetEventFd() const
		{
			return this->efd;
		}
		void ReleaseUserDataEntry(size_t idx)
		{
			this->availableUserDataEntries.push(idx);
		}

	private:
		UserData* GetUserData();

	private:
		// io_uring instance.
		io_uring ring;
		// Event file descriptor to watch for completions.
		int efd;
		// libuv handle used to poll uring completions.
		uv_poll_t* uvHandle{ nullptr };
		// Whether we are currently sending RTP over io_uring.
		bool active{ false };
		// Pre-allocated UserData entries.
		UserData userDataBuffer[QueueDepth]{};
		// Indexes of available UserData entries.
		std::queue<size_t> availableUserDataEntries;
	};
};

#endif
```

ibc commented on worker/src/handles/TcpConnectionHandle.cpp:
```diff
@@ -3,6 +3,7 @@
 
 #include "handles/TcpConnectionHandle.hpp"
 #include "DepLibUV.hpp"
+#include "DepLibUring.hpp"
```
```suggestion
#ifdef MS_LIBURING_SUPPORTED
#include "DepLibUring.hpp"
#endif
```

ibc commented on worker/src/handles/UdpSocketHandle.cpp:
```diff
@@ -2,6 +2,7 @@
 // #define MS_LOG_DEV_LEVEL 3
 
 #include "handles/UdpSocketHandle.hpp"
+#include "DepLibUring.hpp"
```
```suggestion
#ifdef MS_LIBURING_SUPPORTED
#include "DepLibUring.hpp"
#endif
```

ibc commented:
Sorry for so many comments, I thought I was creating a batch but it was not true.

ibc committed:
Merge branch 'v3' into liburing

ibc committed:
Merge branch 'v3' into liburing

jmillan committed:
use MS_LIBURING_SUPPORTED macro everywhere needed

jmillan committed:
Perf: add two optiomisation flags

jmillan committed:
remove flag

Due to its downsides defined here:

https://github.com/axboe/liburing/wiki/io_uring-and-networking-in-2023

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,417 @@
+
+inline static void onFdEvent(uv_poll_t* handle, int status, int events)
+{
+	auto* liburing = static_cast<DepLibUring::LibUring*>(handle->data);
+
+	auto count = io_uring_peek_batch_cqe(liburing->GetRing(), cqes, DepLibUring::QueueDepth);
```
```suggestion
	auto* liburing = static_cast<DepLibUring::LibUring*>(handle->data);
	auto count     = io_uring_peek_batch_cqe(liburing->GetRing(), cqes, DepLibUring::QueueDepth);
```

ibc commented on worker/src/handles/TcpConnectionHandle.cpp:
```diff
@@ -232,6 +235,29 @@ void TcpConnectionHandle::Write(
 	}
 
+#ifdef MS_LIBURING_SUPPORTED
+	{
+		if (!DepLibUring::IsActive())
+		{
+			goto write_libuv;
+		}
+
+		// Prepare the data to be sent.
+		// NOTE: If all SQEs are currently in use or no UserData entry is available we'll
+		// fall back to libuv.
+		auto prepared = DepLibUring::PrepareWrite(this->fd, data1, len1, data2, len2, cb);
+
+		if (!prepared)
+		{
+			// Cannot write via liburing, fallback to libuv.
+			goto write_libuv;
+		}
+
+		return;
+	}
```
We don't need those brackets.

```suggestion
	if (!DepLibUring::IsActive())
	{
		goto write_libuv;
	}

	// Prepare the data to be sent.
	// NOTE: If all SQEs are currently in use or no UserData entry is available we'll
	// fall back to libuv.
	auto prepared = DepLibUring::PrepareWrite(this->fd, data1, len1, data2, len2, cb);

	if (!prepared)
	{
		// Cannot write via liburing, fallback to libuv.
		goto write_libuv;
	}

	return;
```

jmillan replied:
The brackets are there for a reason. Try to remove them and it won't compile:

<img width="1389" alt="Screenshot 2023-11-20 at 16 58 35" src="https://github.com/versatica/mediasoup/assets/732340/1968823e-ece1-4003-98ba-b83aed63f4c3">

ibc replied:
Hell, ok.

ibc commented on worker/src/handles/UdpSocketHandle.cpp:
```diff
@@ -157,6 +160,29 @@ void UdpSocketHandle::Send(
+
+		if (!prepared)
+		{
+			// Cannot send via liburing, fallback to libuv.
```
Some MS_DEBUG_DEV saying it?

ibc commented on worker/src/handles/TcpConnectionHandle.cpp:
```diff
@@ -232,6 +235,29 @@ void TcpConnectionHandle::Write(
+
+		if (!prepared)
+		{
+			// Cannot write via liburing, fallback to libuv.
```
Some MS_DEBUG_DEV saying it?

ibc commented on worker/src/handles/UdpSocketHandle.cpp:
```diff
@@ -322,6 +348,13 @@ bool UdpSocketHandle::SetLocalAddress()
+
+	if (err != 0)
+	{
+		MS_ERROR("uv_fileno() failed: %s", uv_strerror(err));
```
Why this? When can this fail? And what would happen if this fail? How is this related to liburing? Also, in p TcpSocketHandle.cpp` here you are also calling `return false;` but here that0s missing.

jmillan replied:
> Why this? When can this fail? 

The method I'm calling to retrieve the file descriptor can fail, so we must handle it. I should not fail but it can.

> And what would happen if this fail?

We should abort, I've added a `return false` right now.

EDIT: The calling method will abort if this method returns `false`.

> How is this related to liburing? 

libuv does know the file descriptor of its handles, but in liburing we need to tell it which file descriptor we want to send data on.

> TcpSocketHandle.cpphere you are also callingreturn false;` but here that0s missing.

Fixed.

ibc replied:
> > And what would happen if this fail?
> 
> We should abort, I've added a `return false` right now.

But what are the implications of returning false? Or should we replace with MS_ABORT()?


> > How is this related to liburing?
> 
> libuv does know the file descriptor of it's handles, but in liburing we need to tell it which file descriptor we want to send data on.

I don't get it. I mean, I do, but why is this now needed? Literally in `UdpSocketHandle` the PR does this:

```c++
	Utils::IP::GetAddressInfo(
	  reinterpret_cast<const struct sockaddr*>(&this->localAddr), family, this->localIp, this->localPort);


	// This is new in this PR:
	err = uv_fileno(reinterpret_cast<uv_handle_t*>(this->uvHandle), std::addressof(this->fd));

	if (err != 0)
	{
		MS_ERROR("uv_fileno() failed: %s", uv_strerror(err));
```

Why isn't that protected by `#ifdef MS_LIBURING_SUPPORTED` macro?

ibc replied:
And why is that code in `SetLocalAddress()` method? That method is only supposed to obtain local IP and port in string/integer format, nothing else.

jmillan replied:
> But what are the implications of returning false? Or should we replace with MS_ABORT()?

If `this.SetLocalAddress()` returns false, then the calling code will already call MS_ABORT.

>  but why is this now needed? 

We need the file descriptor of our local sockets in order to tell to liburing to send data over those file descriptors.

> Why isn't that protected by #ifdef MS_LIBURING_SUPPORTED macro?

I can put the macro sure, and also to the definition of `this->fd`. Do you want it?

> And why is that code in SetLocalAddress() method? That method is only supposed to obtain local IP and port in string/integer format, nothing else.

Let me put out of there.

jmillan replied:
Handled,

- Added the definition and usage of `this->fd` inside the macro.
- Moved the retrieval of `this->fd` out of those methods.

jmillan committed:
format

jmillan committed:
add debug logs

jmillan committed:
fail if we cannot get the local file descriptor

jmillan committed:
fix comment

jmillan committed:
feedback: use MS_LIBURING_SUPPORTED around

jmillan committed:
add comments

ibc reviewed (commented):
Love it now. Is it ready? Are you gonna add those entries we discussed into the dump data? Or in another PR?

jmillan commented:
> Love it now. Is it ready? Are you gonna add those entries we discussed into the dump data? Or in another PR?

Thanks, I'll do few changes in this PR and add those dump data too. I'll notify you once done.

jmillan committed:
use liburing to send Nacked RTP

RTP packets are retransmitted in batch, determined by the NACK RTCP
packet. Make use of a single system call to send all Nacked packets.

jmillan committed:
use liburing to send RTCP compound packets

Take advantage of a single system call when sending multiple compound
packets.

jmillan committed:
rename DepLibUring::IsSupported by IsRuntimeSupported

The macro MS_LIBURING_SUPPORTED indicates that liburing capabilities are
contained in the binary.

IsRuntimeSupported() indicates whether the runtime can make use of such
capabilities.

jmillan committed:
Liburing::Dump()

jmillan commented:
@ibc, the initial version of liburing is ready to merge. Can you please take a look at today's 4 commits?

* Send Nack-ed RTP using liburing
* Send RTCP compound packets using liburing
* Rename `IsSupported()` method for `IsRuntimeSupported()`
* `LibUring::Dump()`

ibc committed:
Merge branch 'v3' into liburing

ibc commented on worker/include/DepLibUring.hpp:
```diff
@@ -0,0 +1,102 @@
+		// Indexes of available UserData entries.
+		std::queue<size_t> availableUserDataEntries;
+		// Submission queue entry process count.
+		uint64_t sqeProcessCount{ 0 };
+		// Submission queue entry miss count.
+		uint64_t sqeMissCount{ 0 };
+		// User data miss count.
+		uint64_t userDataMissCount{ 0 };
```
Consistency.

```suggestion
		uint64_t sqeProcessCount{ 0u };
		// Submission queue entry miss count.
		uint64_t sqeMissCount{ 0u };
		// User data miss count.
		uint64_t userDataMissCount{ 0u };
```

ibc commented on worker/src/DepLibUring.cpp:
```diff
@@ -0,0 +1,439 @@
+{
+	MS_TRACE();
+
+	MS_DUMP("<LibUring>");
+	MS_DUMP("  sqeProcessCount        : %" PRIu64, this->sqeProcessCount);
+	MS_DUMP("  sqeMissCount           : %" PRIu64, this->sqeMissCount);
+	MS_DUMP("  userDataMissCount      : %" PRIu64, this->sqeMissCount);
+	MS_DUMP("</LibUring>");
```
```suggestion
	MS_DUMP("<DepLibUring>");
	MS_DUMP("  sqeProcessCount   : %" PRIu64, this->sqeProcessCount);
	MS_DUMP("  sqeMissCount      : %" PRIu64, this->sqeMissCount);
	MS_DUMP("  userDataMissCount : %" PRIu64, this->sqeMissCount);
	MS_DUMP("</DepLibUring>");
```

But why do we need this? This is not useful. I expected this info in the worker dump (meaning in the FBS dump response).

jmillan replied:
It's just a mock method. The important thing is that it holds the values.

OK, I'll add this to the worker dump data.

jmillan replied:
It's `LibUring` class where we are doing the `Dump()`.

jmillan replied:
I'm removing this anyway.

jmillan committed:
consistency in numerical initializations

jmillan committed:
fix Dump()

jmillan committed:
FBS: LibUring::Dump

jmillan committed:
Add missing liburing.fbs

jmillan committed:
format

jmillan committed:
Node: lint

jmillan committed:
Node: adapt tests for 'worker.dump()'

jmillan committed:
Update CHANGELOG.md

jmillan committed:
Update CHANGELOG.md

ashinrot commented:
> This greately reduces the number of system calls as just a single system call is needed to relay the incoming RTP packet to "every" Consumer

Hello, I'm working on udp syscall tunning on my local server, and trying lots of linux feature. IOUring seems to be a good solution, but how about `sendmmsg` syscall, it seems to be able to reduce the syscall with less code?

jmillan commented:
> but how about sendmmsg syscall, it seems to be able to reduce the syscall with less code?

You don't send multiple RTP packets in a raw to the same socket in mediasoup, but the "same" RTP packet from a Producer, to multiple sockets.

`sendmmsg` is for sending multiple packets to the same socket.

This PR is about using `liburing` when it's more convenient than `libuv`.
